packages/autoloader/src/class-plugins-handler.php

```diff
@@ -0,0 +1,125 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * This class provides information about the current plugin and the site's active plugins.
+ */
+class Plugins_Handler {
+
+	/**
+	 * Returns an array containing the active plugins. If any plugins are activating,
+	 * they are included in the array.
+	 *
+	 * @return Array An array of plugin names as strings.
+	 */
+	public function get_active_plugins() {
+		$active_plugins = array_merge(
+			is_multisite()
+				? array_keys( get_site_option( 'active_sitewide_plugins', array() ) )
+				: array(),
+			(array) get_option( 'active_plugins', array() )
+		);
+		$current_plugin = $this->get_current_plugin();
+
+		if ( ! in_array( $current_plugin, $active_plugins, true ) ) {
+			// The current plugin isn't active, so it must be activating. Add it to the list.
+			$active_plugins[] = $current_plugin;
+		}
+
+		// If the activating plugin is not the only activating plugin, we need to add others too.
+		$active_plugins = array_unique( array_merge( $active_plugins, $this->get_activating_plugins() ) );
+
+		return $active_plugins;
+	}
+
+	/**
+	 * Creates an array containing the paths to the classmap and filemap for the given plugin.
+	 * The classmap and filemap filenames are the names of the files generated by Jetpack
+	 * Autoloader with versions >=2.0.
+	 *
+	 * @param String $plugin The plugin string.
+	 * @return Array An array containing the paths to the plugin's classmap and filemap.
+	 */
+	public function create_map_path_array( $plugin ) {
+		$plugin_path = plugin_dir_path( trailingslashit( WP_PLUGIN_DIR ) . $plugin );
+
+		return array(
+			'class' => trailingslashit( $plugin_path ) . 'vendor/composer/jetpack_autoload_classmap.php',
+			'file'  => trailingslashit( $plugin_path ) . 'vendor/composer/jetpack_autoload_filemap.php',
+		);
+	}
+
+	/**
+	 * Returns an array containing the paths to the classmap and filemap for the active plugins.
+	 */
+	public function get_active_plugins_paths() {
+		$active_plugins = $this->get_active_plugins();
+		return array_map( array( $this, 'create_map_path_array' ), $active_plugins );
+	}
+
+	/**
+	 * Checks whether the current plugin is active.
+	 *
+	 * @return Boolean True if the current plugin is active, else False.
+	 */
+	public function is_current_plugin_active() {
+		$active_plugins = (array) get_option( 'active_plugins', array() );
+		$current_plugin = $this->get_current_plugin();
+
+		return in_array( $current_plugin, $active_plugins, true );
+	}
```

Comment:
I think we might want to return True if the plugin is active or if it's a known activating plugin (see `get_activating_plugins()` ). `is_current_plugin_active()` is used in `set_up_autoloader()` to determine if the autoloader globals should be reset. If we already know that the plugin is activating, the globals shouldn't be reset.

Reply:
If I understand your idea here correctly, you mean that we should look for the current plugin in the following list:
```
	$active_plugins = $this->get_active_plugins();
```

Right now we're doing this instead:
```
	$active_plugins = (array) get_option( 'active_plugins', array() );
```

If we use `get_active_plugins`, we will return true in `is_current_plugin_active` every time that we can detect the plugin activation, so the autoloader globals reset will be reserved only for the case when we can't detect its activation, right?

If that's so, I think that should work fine.

Reply:
Yes, that's exactly what I was thinking. I made this change in commit 612e84f.

Reply:
I forgot that `get_active_plugins()` added the current plugin to the active plugins list if it wasn't already there. So, I couldn't use that method to determine whether to reset the autoloader.

I pushed a commit that refactors the `Plugins_Handler` class a bit. With the changes in that commit, the autoloader will only reset when the plugin was unknown (i.e. not an active plugin and not an already known activating plugin).